### PR TITLE
Column settings, mini bar charts, and more

### PR DESCRIPTION
**Pull request by tlrobinson**
* [x] remove `comma` formatting option
* [x] merge Scalar settings into formatting options
* [x] hide "Markdown template" setting
  * [x] bring back prefix/suffix settings?
* [x] resize columns when changing format settings like mini bar chart
* [ ] empty state / what to do if there aren't settings for a particular column?
* [x] consolidate "formatting" and "data" for tables
* [x] date_format currently only for `day` unit or no unit, what about `month`, `week`, etc?
* [x] links + images settings
*  ~enable for all visualization types~ follow-on PR
- [x] seems like the Quarter granularity style’s radio button doesn’t have a default set
- ~still need to disable settings for columns that don’t have any~
- [x] Is “column title” a setting we should put in? So you could change e.g. “Sum” to “Revenue”. That might also have the side effect of making every column have at least one setting. (scope creep, though)
- [ ] does showing a mini bar chart for lat/long ever make sense? it’s in the settings for fields of that type right now.
- [x] you’re right that for dates we should probably do 1/7/2018 (MM-DD-YYYY)

### CODE REVIEW

**Comment by mazameli** [comment]
I couldn't get the third or fourth separator style settings to take effect, either with a SQL query, or with a GUI one.

separators

**Comment by tlrobinson** [comment]
@mazameli pushed a fix for that.

**Comment by mazameli** [comment]
It looks like there's a maximum number of decimal places you can specify (10). Instead of an input box, should we use a stepped range slider from 0 to 10, defaulting to 2? Not sure if this is in scope for this particular PR.

**Comment by tlrobinson** [comment]
The max is 20 so I'm not too worried about people going over. But yeah we could do something like that, maybe as part of cleaning up / expanding the scalar settings screen though.

**Comment by mazameli** [comment]
Solid progress. Some notes after playing with it a bit just now:

- We should talk about the markdown feature. Right now it's the only option displayed when you click on the options for any field other than a number or date, and it doesn't feel right to me to be displaying a power-user feature like this so (relatively) prominently. It's also gonna need some explaining.
- In the scalar settings, it feels funny to me to have the number of decimal places input box be empty instead of showing `2`, which is what the actual setting default is.
- As time allows, we could add some more basic settings to other field types. E.g. URL fields could have a Link Text option to display a string instead of the URL, and email addresses could have a toggle for whether or not they should be clickable (kind of annoying to have Mail.app open up when you're just trying to copy an email address).
- We'll want to polish the list of fields together tomorrow, and for tables, look at possibly combining the `formatting` and `data` tabs into a `fields` tab since they feel somewhat redundant now.

**Comment by mazameli** [comment]
Also, I love the little mini bar chart. I did notice there are cases like this where it looks squashed by default unless I manually expand the column width:

screen shot 2018-09-13 at 4 50 14 pm

**Comment by tlrobinson** [comment]
> In the scalar settings, it feels funny to me to have the number of decimal places input box be empty instead of showing 2, which is what the actual setting default is.

That's not actually the default since that setting is for an *exact* number of decimal places, not the maximum number of decimal places. `1` will be displayed as `1` if it's not set but `1.00` if it is set. I think we added that to fake currency formatting but maybe we can get rid of it, or replace it with something that behaves like what you're talking about ("precision"? "maximum decimal places"?)

**Comment by tlrobinson** [comment]
> Also, I love the little mini bar chart. I did notice there are cases like this where it looks squashed by default unless I manually expand the column width:

I think the problem is that we measure and fix the column width before you have a chance to turn on the mini bar. I can probably come up with a way to reset it when you toggle that setting.

**Comment by mazameli** [comment]
> That's not actually the default since that setting is for an exact number of decimal places, not the maximum number of decimal places. 1 will be displayed as 1 if it's not set but 1.00 if it is set. I think we added that to fake currency formatting but maybe we can get rid of it, or replace it with something that behaves like what you're talking about ("precision"? "maximum decimal places"?)

Ah yeah, duh, brain fart. We could have it be a toggle next to the input box where you could switch between `Exactly` and `Maximum` or something.

> I think the problem is that we measure and fix the column width before you have a chance to turn on the mini bar. I can probably come up with a way to reset it when you toggle that setting.

Cool. I noticed a similar (I think) problem when I had the bars turned on, and then I set that same column to currency: the addition of the currency symbol caused the values to collide with the bars.

**Comment by tlrobinson** [comment]
> We should talk about the markdown feature. Right now it's the only option displayed when you click on the options for any field other than a number or date, and it doesn't feel right to me to be displaying a power-user feature like this so (relatively) prominently. It's also gonna need some explaining.

Yeah, I think it makes sense to provide more "structured" settings (that perhaps generate a markdown template under the hood) where possible. Maybe it only makes sense to show for columns that don't have a `special_type` (e.x. native queries, calculated columns), or possibly better, allow the user to select a `special_type` for those columns so they can get the richer formatting settings.

**Review by kdoh — COMMENTED** [review]
Left a note about some extremely minor quibbles. In general from the implementation side I think this is a nice change since it makes it a littler easier to just go check the column settings file to get a better sense of what the possible settings are for the type of data you're visualizing. 

Couple things I reacted to while reading through the diff that are more points of discussion rather than good or bad things.

1. I'm a little nervous about the lack of unit / component tests since the surface area of this is so large. Ultimately I think it makes our life easier because we're condensing the places the logic can live but I'd still feel a little better if we had a smidge more coverage even if its just for some of the possible outputs of `columnSettings`

2. I still feel like the idiosyncrasies of why we need to do certain manipulations on series or check if something is a scatter chart are still primarily living in Tom's Brain™ if that makes sense. This isn't a new thing for this diff but as I was reading through it I still felt like I don't necessarily understand the purpose for some of the functions other than "they end up making things look right." Some simple comments on some of them about what they do and why would make it a little easier to understand the system as a whole I think. For me `getComputedSettings` and its cousin `getPersistableDefaultSettings` are examples where if asked I couldn't tell you why or when to use one vs the other, or if they should be used in concert, other than a general sense of one is computed and one is default. Ultimately there's enough complexity to this that a run down of the flow that wasn't just the code itself is probably helpful to have somewhere as a reference.

Conversely as a result of this diff I feel like I *do* understand the range of possible settings for different column types and how to add new ones, which I didn't as much before, so it'd be nice if we moved some of the underlying logic in a similar positive direction.

File: frontend/src/metabase/visualizations/lib/settings/utils.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import { isDimension, isMetric } from "metabase/lib/schema_metadata";
+import { getFriendlyName } from "metabase/visualizations/lib/utils";
+
+export const DIMENSION_METRIC = "DIMENSION_METRIC";
+export const DIMENSION_METRIC_METRIC = "DIMENSION_METRIC_METRIC";
```
[the remark this answers is not in the document]
Comment (tlrobinson) [reply frontend/src/metabase/visualizations/lib/settings/utils.js]: This is old code that I just moved around.

I suppose we could rename it to `SINGLE_DIMENSION_MULTIPLE_METRIC` or something.

**Pull request merged**